GoogleCloudPlatform/k8s-config-connector
# Graduate customization CRDs to v1beta1

### Change description


 - Add v1beta1 API, which is the same as the v1alpha1.
 - Mark v1beta1 as the storage version.
 - Mark v1alpha1 as not served version.

### Tests you have done



1. A fresh install of KCC v1.1110.0 (which has v1alpha1 customziation CRD)

```
$ git checkout v1.110.0

$ cd operator

$ git checkout upstream/master Makefile
(For convenience, we need newly added Makefile rules from HEAD)

$ make uninstall destroy

$ make install docker-build docker-push deploy
```

2. Create a CC object
```
$ k apply -f cc.yaml 
configconnector.core.cnrm.cloud.google.com/configconnector.core.cnrm.cloud.google.com created
```

3. Create v1alpha1 customization CRs
```
$ k apply -f config/samples/controller_resource_customization_sample.yaml 
controllerresource.customize.core.cnrm.cloud.google.com/cnrm-controller-manager created
controllerresource.customize.core.cnrm.cloud.google.com/cnrm-webhook-manager created

$ k get controllerresources
NAME                      AGE
cnrm-controller-manager   9s
cnrm-webhook-manager      9s
```

4. Verify the customization is applied
```
$ k get deployment cnrm-webhook-manager -n cnrm-system -o jsonpath='{range .spec.template.spec.containers[*]}[{.name}, {.resources}]{"\n"}{end}' 
[webhook, map[limits:map[memory:512Mi] requests:map[cpu:250m memory:256Mi]]]

$ k get pods -n cnrm-system (4 replicas for webhook manager pod)
NAME                                           READY   STATUS    RESTARTS   AGE
cnrm-controller-manager-0                      2/2     Running   0          20s
cnrm-deletiondefender-0                        1/1     Running   0          107s
cnrm-resource-stats-recorder-b766c4fd6-nskvr   2/2     Running   0          108s
cnrm-webhook-manager-d5887dc87-6npkj           1/1     Running   0          19s
cnrm-webhook-manager-d5887dc87-7wb4b           1/1     Running   0          11s
cnrm-webhook-manager-d5887dc87-t5gwz           1/1     Running   0          31s
cnrm-webhook-manager-d5887dc87-wpk92           1/1     Running   0          31s
```

5. Install operator and CRDs from this PR (This step mimics the KCC version upgrade)
```
$ git checkout v1beta1_customization_CRDs

$ make install docker-build docker-push deploy
```

6. Verify the previously created customization CRs still exist
```
$ k get controllerresources
Error from server (NotFound): Unable to list "customize.core.cnrm.cloud.google.com/v1alpha1, Resource=controllerresources": the server could not find the requested resource (get controllerresources.customize.core.cnrm.cloud.google.com)

$ k get controllerresources.customize.core.cnrm.cloud.google.com
NAME                      AGE
cnrm-controller-manager   4m46s
cnrm-webhook-manager      4m46s
```

7. Verify the customization is still effective
```
$ k get deployment cnrm-webhook-manager -n cnrm-system -o jsonpath='{range .spec.template.spec.containers[*]}[{.name}, {.resources}]{"\n"}{end}' 
[webhook, map[limits:map[memory:512Mi] requests:map[cpu:250m memory:256Mi]]]

$ k get pods -n cnrm-system
NAME                                           READY   STATUS    RESTARTS   AGE
cnrm-controller-manager-0                      2/2     Running   0          4m41s
cnrm-deletiondefender-0                        1/1     Running   0          6m8s
cnrm-resource-stats-recorder-b766c4fd6-nskvr   2/2     Running   0          6m9s
cnrm-webhook-manager-d5887dc87-6npkj           1/1     Running   0          4m40s
cnrm-webhook-manager-d5887dc87-7wb4b           1/1     Running   0          4m32s
cnrm-webhook-manager-d5887dc87-t5gwz           1/1     Running   0          4m52s
cnrm-webhook-manager-d5887dc87-wpk92           1/1     Running   0          4m52s
``` 

8. The existing CRs can be read via v1beta1
```
$ k get -f config/samples/controller_resource_customization_sample.yaml
NAME                      AGE
cnrm-controller-manager   7m7s
cnrm-webhook-manager      7m7s
```

9. v1alpha1 is no longer served
```
$ git checkout v1.110.0

$ k get -f config/samples/controller_resource_customization_sample.yaml
unable to recognize "config/samples/controller_resource_customization_sample.yaml": no matches for kind "ControllerResource" in version "customize.core.cnrm.cloud.google.com/v1alpha1"
unable to recognize "config/samples/controller_resource_customization_sample.yaml": no matches for kind "ControllerResource" in version "customize.core.cnrm.cloud.google.com/v1alpha1"
```
